web/modules/custom/dpl_login/src/UserTokenAuthProvider.php

```diff
@@ -0,0 +1,75 @@
+<?php
+
+namespace Drupal\dpl_login;
+
+use Drupal\Core\Authentication\AuthenticationProviderInterface;
+use Drupal\Core\Extension\ModuleHandlerInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\openid_connect\OpenIDConnectAuthmap;
+use Drupal\openid_connect\Plugin\OpenIDConnectClientInterface;
+use Symfony\Component\HttpFoundation\Request;
+use function Safe\preg_match as preg_match;
+
+/**
+ * Authentication provider with OAuth2 / OpenID Connect user token support.
+ */
+class UserTokenAuthProvider implements AuthenticationProviderInterface {
+
+  /**
+   * Constructor.
+   */
+  public function __construct(
+    private OpenIDConnectClientInterface $client,
+    private ModuleHandlerInterface $moduleHandler,
+    private OpenIDConnectAuthmap $authmap,
+  ) {}
+
+  /**
+   * Extract a bearer token from the authorization header of a request.
+   */
+  private function getToken(Request $request) : ?string {
+    $header = $request->headers->get('Authorization');
+    if (!$header) {
+      return NULL;
+    }
```

Comment:
It is a matter of taste but I usually solve situations like this:
```php
    if (!$header = $request->headers->get('Authorization')) {
      return NULL;
    }

Reply:
I have noticed and we differ in taste here :smile:

In my opinion assignments in conditions leads to less readable code and risk problems of mixing up assignment and comparison.